docs/sql/json.md

```diff
@@ -0,0 +1,243 @@
+
+Values of type `ARRAY`, `MAP`, and `ROW` type can be cast to `VARIANT`.  `VARIANT` values
+also offer the following operations:
+
+- indexing using array indexing notation `variant[index]`.  If the `VARIANT` is
+  obtained from an `ARRAY` value, the indexing operation returns a `VARIANT` whose value element
+  is the element at the specified index.  Otherwise, this operation returns `NULL`
+- indexing using map element access notation `variant[key]`, where `key` can have
+  any legal `MAP` key type.  If the `VARIANT` is obtained from a `MAP` value
+  that has en element with this key, a `VARIANT` value holding the associated value in
+  the `MAP` is returned.  Otherwise `NULL` is returned.  If the `VARIANT` is obtained from `ROW` value
+  which has a field with the name `key`, this operation returns a `VARIANT` value holding
+  the corresponding field value.  Otherwise `NULL` is returned.
+- field access using the dot notation: `variant.field`.  This operation is interpreted
+  as equivalent to `variant['field']`.  Note, however, that the field notation
+  is subject to the capitalization rules of the SQL dialect, so for correct
+  operation the field may need to be quoted: `variant."field"`
+
+The runtime types do not need to match exactly the compile-time types.
+As a compiler front-end, Calcite does not mandate exactly how the runtime types
+are represented.  Calcite does include one particular implementation in
+Java runtime, which is used for testing.  In this representation
+the runtime types are represented as follows:
+
+- The scalar types do not include information about precision and scale.  Thus all `DECIMAL`
+  compile-time types are represented by a single run-time type.
+- `CHAR(N)` and `VARCHAR` are both represented by a single runtime `VARCHAR` type.
+- `BINARY(N)` and `VARBINARY` are both represented by a single runtime `VARBINARY` type.
+- `FLOAT` and `DOUBLE` are both represented by the same runtime type.
+- All "short interval" types (from days to seconds) are represented by a single type.
+- All "long interval" types (from years to months) are represented by a single type.
+- Generic types such as `INT ARRAY`, `MULTISET`, and `MAP` do carry runtime
+  information about the element types
+- The `ROW` type does have information about all field types (currently not yet supported)
+
+## Functions that operate on `VARIANT` values
+
+| Function | Description |
+| `VARIANTNULL()` | Can be used to create an instance of the `VARIANT` `null` value. |
+| `TYPEOF(` _variant_ `)` | Argument must be a `VARIANT` value.  Returns the runtime type of the value |
+| `PARSE_JSON(` _string_ `)` | Parses a string that represents a JSON value, returns a `VARIANT` object, or `NULL` if parsing fails |
+| `UNPARSE_JSON(` _variant_ `)` | Returns a string that represents the serialization of a `VARIANT` value. If the value cannot be represented as JSON, the result is `NULL` |
+
+Here are some examples using `VARIANT` and `JSON` values:
```

Comment:
I think we should make all of these into examples that we can actually run in our software, e.g., have a SQL code snippet that is a table and then has views with the casts

Reply:
I have these as tests in the compiler source code, I have pasted the code that I use for tests.
But this requires one different program for each example, since inputs and outputs have different types for each test.

Reply:
what about one big snippet with all these in a single program that they can paste and run

Reply:
There are two problems with this:
- the program does not look very informative:
```sql
CREATE VIEW V0 AS SELECT CAST(1 AS VARIANT);
CREATE VIEW V1 AS SELECT TYPEOF(CAST(1 AS VARIANT));
CREATE VIEW V2 AS SELECT CAST(CAST(1 AS TINYINT) AS VARIANT);
-- The runtime knows that this is a TINYINT
CREATE VIEW V3 AS SELECT TYPEOF(CAST(CAST(1 AS TINYINT) AS VARIANT));
CREATE VIEW V4 AS SELECT CAST(CAST(1 AS VARIANT) AS INT);
-- However, you have to use the right type, or you get NULL
CREATE VIEW V5 AS SELECT CAST(CAST(1 AS VARIANT) AS TINYINT);
-- Some VARIANT objects when output receive double quotes
CREATE VIEW V6 AS select CAST('string' as VARIANT);
-- CHAR(3) values are represented as VARCHAR in variants
```
and so on
- the pipeline manager crashes because it doesn't know about VARIANT

Reply:
surely there must be a snippet we can give here that works, otherwise I doubt anyone 
will figure this out by themselves.

Reply:
But this section is a collection of 50 examples. True, these examples cannot be executed as shown, but I don't know what more to do here. And your ad-hoc queries will probably not be able to run these either.

Reply:
we dont have to worry about adhoc queries for this. especially if the type never appears in a table or view as discussed this morning. But please lets have code that actually works, how is a user supposed to figure this out otherwise?